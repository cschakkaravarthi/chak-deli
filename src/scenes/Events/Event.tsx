import React, { FC, useEffect, useRef, useState } from 'react';
import { connect } from 'react-redux';
import get from 'lodash.get';
import createDOMPurify from 'dompurify';
import Col from 'react-bootstrap/Col';
import EventModel from '../../shared/models/Event.model';
import Row from 'react-bootstrap/Row';
import { AppState } from '../../shared/types/genericTypes';
import { Card } from 'umgc_ui_library/lib';
import { Container } from 'react-bootstrap';
import { fetchEventById } from '../../shared/actions/eventActions';
import { usehandleAddToCalendar, normalizeEventData } from '../../utils/customHooks';
import { formatDateTime, formatDateRangeWithTime, formatDateRange } from '../../shared/services/date';
import BSToast from '../../components/toast/Toast';
import { match } from 'react-router';
import { TaxonomyTermModel } from '../../shared/types/contentTypes';
import Error404 from '../Errors/Error404';

type Props = {
  match?: match;
  event?: EventModel;
  eventsList?: EventModel[];
  moreEventsList?: EventModel[];
  fetchEventById?: (a: number | string) => Promise<void>;
};

const mapState = (state: AppState, props: Props): Props => (
  {
    ...props,
    event: state.eventReducers.event,
    eventsList: state.eventReducers.eventsList,
    moreEventsList: state.eventReducers.moreEventsList
  }
);

const actionCreators = { fetchEventById };

export const Event: FC<Props> = props => {
  const { match, event, eventsList, moreEventsList, fetchEventById } = props;

  const [isFetching, setIsFetching] = useState<boolean>(true);

  const urlEventId = parseInt(get(match, 'params.id'), 10);

  // eventsList stores the events on home page
  // while moreEventsList the events in the events scene.
  const eventStore = moreEventsList || eventsList;

  if (!urlEventId) {
    if (!isFetching) {
      return <Error404 />;
    } else {
      return null;
    }
  }

  const shouldFetch = eventStore
    ? eventStore.find(eventItem => eventItem.drupal_id === urlEventId)
    : undefined;

  useEffect(() => {
    if (!shouldFetch) {
      if (fetchEventById) {
        fetchEventById(urlEventId)
          .then(() => {
            setIsFetching(false);
          })
          .catch(() => {
            setIsFetching(false);
          });
      }
    } else {
      setIsFetching(false);
    }
  }, []);

  const eventResolution = shouldFetch || event;

  const toastRef = useRef();

  const handleAddToCalender = (id: string, title: string): void => {
    usehandleAddToCalendar(id, title, toastRef);
  };

  if (!eventResolution) {
    if (!isFetching) {
      return <Error404 />;
    } else {
      return null;
    }
  }

  const getDate = (event: EventModel): string => {
    const limit = 23 * 60 * 60 + 59 * 60; // 23 hours and 59 seconds
    if (event.when_end) {
      const difference = event.when_end - event.when_start;

      if (difference < limit) {
        return formatDateRangeWithTime(event.when_start, event.when_end);
      } else if (difference === limit) {
        return formatDateTime(event.when_start, 'noTZ');
      } else {
        return formatDateRange(event.when_start, event.when_end);
      }
    } else {
      return formatDateTime(event.when_start, 'noTZ');
    }
  };

  return (
    <Container className="content-full">
      <Row>
        <Col md={12} className="p-0 mt-5">
          <BSToast ref={toastRef}/>
          <Card
            type="event"
            title={eventResolution.title}
            variant="full"
            location={eventResolution.where}
            date={getDate(eventResolution)}
            imageUrl={get(eventResolution, 'image_uri.umgc_featured', '')}
            categories={get(
              eventResolution,
              'event_category',
              [] as TaxonomyTermModel[]
            )}
            content={
              <div
                className="position-relative"
                dangerouslySetInnerHTML={{
                  __html: createDOMPurify.sanitize(
                    get(eventResolution, 'body_full', '')
                  )
                }}
              />
            }
            AddToCalenderLink={() => {
              handleAddToCalender(
                eventResolution.drupal_id.toString(),
                eventResolution.title
              );
            }}
            SendInviteEventData = {normalizeEventData(eventResolution)}
          />
        </Col>
      </Row>
    </Container>
  );
};

export default connect(
  mapState,
  actionCreators
)(Event);
