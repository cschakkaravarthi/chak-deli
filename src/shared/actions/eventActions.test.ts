import moxios from 'moxios';
import {
  loadEvents,
  fetchEvents,
  loadEventById,
  fetchEventById,
  loadMoreEvents,
  fetchHomeEvents
} from './eventActions';
import { dummyEvents } from '../models/Event.model';
import { mockedFacets } from '../models/Facet.model';
import getStore from '../services/mockGlobalStore';

describe('eventActions', () => {
  let store: any;
  beforeEach(() => {
    store = getStore();
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('creates a loadEvents action', () => {
    expect(loadEvents(dummyEvents)).toMatchSnapshot();
  });

  it('creates a loadArticleById action', () => {
    expect(loadEventById(dummyEvents[0])).toMatchSnapshot();
  });

  it('creates a loadMoreEvents action', () => {
    const ContentWithFacets = {
      eventsList: dummyEvents,
      facets: mockedFacets
    };
    expect(loadMoreEvents(ContentWithFacets)).toMatchSnapshot();
  });

  it.skip('fetchHomeEvents makes API call', done => {
    const expectedActions = ['LOAD_EVENTS'];

    moxios.stubRequest('content?type=event&limit=3&sort=latest', {
      status: 200,
      responseText: 'OK'
    });

    store
      .dispatch(fetchHomeEvents())
      .then(() => {
        const actionsCalled = store
          .getActions()
          .map((action: any) => action.type);
        expect(actionsCalled).toEqual(expectedActions);
        done();
      })
      .catch((error: any) => {
        done(error);
      });
  });

  it.skip('fetchEventById makes API call', done => {
    const expectedActions = ['LOAD_EVENT_BY_ID'];

    moxios.stubRequest('events?id=4', {
      status: 200,
      responseText: 'OK'
    });

    store
      .dispatch(fetchEventById('4'))
      .then(() => {
        const actionsCalled = store
          .getActions()
          .map((action: any) => action.type);
        expect(actionsCalled).toEqual(expectedActions);
        done();
      })
      .catch((error: any) => {
        done(error);
      });
  });

  it.skip('fetchEvents makes API call', done => {
    const expectedActions = ['LOAD_MORE_EVENTS'];

    moxios.stubRequest('content?type=event&page=1&limit=5', {
      status: 200,
      responseText: 'OK'
    });

    store
      .dispatch(fetchEvents())
      .then(() => {
        const actionsCalled = store
          .getActions()
          .map((action: any) => action.type);
        expect(actionsCalled).toEqual(expectedActions);
        done();
      })
      .catch((error: any) => {
        done(error);
      });
  });
});
