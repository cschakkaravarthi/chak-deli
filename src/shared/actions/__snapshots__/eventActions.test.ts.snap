// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eventActions creates a loadArticleById action 1`] = `
Object {
  "event": Object {
    "attachments": null,
    "body_full": "<p>Some body text</p>
",
    "categories": Array [],
    "category_ancestors": Array [],
    "created": 1568326940,
    "drupal_id": 4,
    "event_category": Array [
      Object {
        "drupal_id": 46,
        "icon": null,
        "title": "Industry Events",
        "weight": 1,
      },
    ],
    "image_uri": Object {
      "source": "undefined",
      "umgc_banner": "undefined",
      "umgc_embedded": "undefined",
      "umgc_featured": "undefined",
      "umgc_thumbnail": "undefined",
    },
    "language": "en",
    "owner": Array [
      Object {
        "drupal_id": 3,
        "icon": null,
        "title": "Global Travel",
        "weight": 0,
      },
    ],
    "summary": "",
    "tags": Array [
      Object {
        "drupal_id": 6,
        "icon": null,
        "title": "Lady Gaga",
        "weight": 0,
      },
    ],
    "tags_ancestors": Array [
      Object {
        "drupal_id": 6,
        "icon": null,
        "title": "Lady Gaga",
        "weight": 0,
      },
    ],
    "title": "Test 1",
    "type": "event",
    "when_duration": 7199,
    "when_end": 1568370600,
    "when_start": 1568326940,
    "where": "Over the Rainbow",
  },
  "type": "LOAD_EVENT_BY_ID",
}
`;

exports[`eventActions creates a loadEvents action 1`] = `
Object {
  "eventsList": Array [
    Object {
      "attachments": null,
      "body_full": "<p>Some body text</p>
",
      "categories": Array [],
      "category_ancestors": Array [],
      "created": 1568326940,
      "drupal_id": 4,
      "event_category": Array [
        Object {
          "drupal_id": 46,
          "icon": null,
          "title": "Industry Events",
          "weight": 1,
        },
      ],
      "image_uri": Object {
        "source": "undefined",
        "umgc_banner": "undefined",
        "umgc_embedded": "undefined",
        "umgc_featured": "undefined",
        "umgc_thumbnail": "undefined",
      },
      "language": "en",
      "owner": Array [
        Object {
          "drupal_id": 3,
          "icon": null,
          "title": "Global Travel",
          "weight": 0,
        },
      ],
      "summary": "",
      "tags": Array [
        Object {
          "drupal_id": 6,
          "icon": null,
          "title": "Lady Gaga",
          "weight": 0,
        },
      ],
      "tags_ancestors": Array [
        Object {
          "drupal_id": 6,
          "icon": null,
          "title": "Lady Gaga",
          "weight": 0,
        },
      ],
      "title": "Test 1",
      "type": "event",
      "when_duration": 7199,
      "when_end": 1568370600,
      "when_start": 1568326940,
      "where": "Over the Rainbow",
    },
    Object {
      "attachments": null,
      "body_full": "<p>Some body text</p>
",
      "categories": Array [],
      "category_ancestors": Array [],
      "created": 1568326940,
      "drupal_id": 24,
      "event_category": Array [
        Object {
          "drupal_id": 46,
          "icon": null,
          "title": "Industry Events",
          "weight": 1,
        },
      ],
      "image_uri": Object {
        "source": "undefined",
        "umgc_banner": "undefined",
        "umgc_embedded": "undefined",
        "umgc_featured": "undefined",
        "umgc_thumbnail": "undefined",
      },
      "language": "en",
      "owner": Array [
        Object {
          "drupal_id": 3,
          "icon": null,
          "title": "Global Travel",
          "weight": 0,
        },
      ],
      "summary": "",
      "tags": Array [
        Object {
          "drupal_id": 6,
          "icon": null,
          "title": "Lady Gaga",
          "weight": 0,
        },
      ],
      "tags_ancestors": Array [
        Object {
          "drupal_id": 6,
          "icon": null,
          "title": "Lady Gaga",
          "weight": 0,
        },
      ],
      "title": "Test 1",
      "type": "event",
      "when_duration": 7199,
      "when_end": 1568370600,
      "when_start": 1568326940,
      "where": "Over the Rainbow",
    },
  ],
  "type": "LOAD_EVENTS",
}
`;

exports[`eventActions creates a loadMoreEvents action 1`] = `
Object {
  "facets": Object {
    "tags": Array [
      Object {
        "drupal_id": 1,
        "icon": "string",
        "title": "string",
        "weight": 1,
      },
    ],
  },
  "moreEventsList": Array [
    Object {
      "attachments": null,
      "body_full": "<p>Some body text</p>
",
      "categories": Array [],
      "category_ancestors": Array [],
      "created": 1568326940,
      "drupal_id": 4,
      "event_category": Array [
        Object {
          "drupal_id": 46,
          "icon": null,
          "title": "Industry Events",
          "weight": 1,
        },
      ],
      "image_uri": Object {
        "source": "undefined",
        "umgc_banner": "undefined",
        "umgc_embedded": "undefined",
        "umgc_featured": "undefined",
        "umgc_thumbnail": "undefined",
      },
      "language": "en",
      "owner": Array [
        Object {
          "drupal_id": 3,
          "icon": null,
          "title": "Global Travel",
          "weight": 0,
        },
      ],
      "summary": "",
      "tags": Array [
        Object {
          "drupal_id": 6,
          "icon": null,
          "title": "Lady Gaga",
          "weight": 0,
        },
      ],
      "tags_ancestors": Array [
        Object {
          "drupal_id": 6,
          "icon": null,
          "title": "Lady Gaga",
          "weight": 0,
        },
      ],
      "title": "Test 1",
      "type": "event",
      "when_duration": 7199,
      "when_end": 1568370600,
      "when_start": 1568326940,
      "where": "Over the Rainbow",
    },
    Object {
      "attachments": null,
      "body_full": "<p>Some body text</p>
",
      "categories": Array [],
      "category_ancestors": Array [],
      "created": 1568326940,
      "drupal_id": 24,
      "event_category": Array [
        Object {
          "drupal_id": 46,
          "icon": null,
          "title": "Industry Events",
          "weight": 1,
        },
      ],
      "image_uri": Object {
        "source": "undefined",
        "umgc_banner": "undefined",
        "umgc_embedded": "undefined",
        "umgc_featured": "undefined",
        "umgc_thumbnail": "undefined",
      },
      "language": "en",
      "owner": Array [
        Object {
          "drupal_id": 3,
          "icon": null,
          "title": "Global Travel",
          "weight": 0,
        },
      ],
      "summary": "",
      "tags": Array [
        Object {
          "drupal_id": 6,
          "icon": null,
          "title": "Lady Gaga",
          "weight": 0,
        },
      ],
      "tags_ancestors": Array [
        Object {
          "drupal_id": 6,
          "icon": null,
          "title": "Lady Gaga",
          "weight": 0,
        },
      ],
      "title": "Test 1",
      "type": "event",
      "when_duration": 7199,
      "when_end": 1568370600,
      "when_start": 1568326940,
      "where": "Over the Rainbow",
    },
  ],
  "type": "FETCH_MORE_EVENTS",
}
`;
